In [1]:
import pandas as pd

from arrow import now
from charset_normalizer import detect
from glob import glob

from warnings import filterwarnings
filterwarnings(action='ignore', category=FutureWarning)

CODES = '/kaggle/input/country-mapping-iso-continent-region/continents2.csv'
CROPS = '/kaggle/input/crop-production/P*.csv'

def get_file(filename: str):
    with open(file=filename, mode='rb') as input_fp:
        encoding = detect(input_fp.read())['encoding']
        return pd.read_csv(filepath_or_buffer=filename, encoding=encoding)

time_start = now()
df = pd.concat(objs=[get_file(filename=input_file) for input_file in glob(CROPS)]).merge(right=pd.read_csv(filepath_or_buffer=CODES, usecols=['alpha-3', 'name']), left_on='Area', right_on='name')
print('data read complete in {}'.format(now() - time_start))

df.head()

data read complete in 0:00:12.832777


Area Code       Area  Item Code                              Item  \
0         10  Australia        221               Almonds, with shell   
1         10  Australia        221               Almonds, with shell   
2         10  Australia        221               Almonds, with shell   
3         10  Australia        711  Anise, badian, fennel, coriander   
4         10  Australia        711  Anise, badian, fennel, coriander   

   Element Code         Element    Unit  Y1961 Y1961F  Y1962  ...     Y2016  \
0          5312  Area harvested      ha    NaN      M    NaN  ...   28553.0   
1          5419           Yield   hg/ha    NaN    NaN    NaN  ...   42050.0   
2          5510      Production  tonnes    NaN      M    NaN  ...  120065.0   
3          5312  Area harvested      ha    NaN      M    NaN  ...    1083.0   
4          5419           Yield   hg/ha    NaN    NaN    NaN  ...   11625.0   

   Y2016F     Y2017  Y2017F     Y2018  Y2018F     Y2019  Y2019F       name  \
0     NaN   29282.0     NaN   31361.0     NaN   37903.0     NaN  Australia   
1      Fc   40909.0      Fc   39769.0      Fc   38628.0      Fc  Australia   
2      Im  119790.0      Im  124718.0      Im  146410.0      Im  Australia   
3      Im    1108.0      Im    1145.0      Im    1182.0      Im  Australia   
4      Fc   11661.0      Fc   11712.0      Fc   11751.0      Fc  Australia   

   alpha-3  
0      AUS  
1      AUS  
2      AUS  
3      AUS  
4      AUS  

[5 rows x 127 columns]

In [2]:
df.info() # we have lost 210 - 178 areas in the ISO-3 code merge

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31434 entries, 0 to 31433
Columns: 127 entries, Area Code to alpha-3
dtypes: float64(59), int64(3), object(65)
memory usage: 30.5+ MB


In [3]:
print('we have data for {} areas, {} crops, and {} years.'.format(df['Area'].nunique(), df['Item Code'].nunique(), len([column for column in df.columns if column.startswith('Y')]) // 2))

we have data for 178 areas, 174 crops, and 59 years.


What do when we want to visualize a single series of named data and we have too much to make a bar plot? 

We can introduce a random variable to add a dimension and make a scatter plot, or we can introduce two random variables and lay out the data at random points in our 2-d space. Unfortunately in this case our item codes bunch up, so we need to use two random variables; below we've chosen to use text instead of hover text to show the item names; the result is middling, but it gives us a way to visually browse 174 items and their codes.

In [4]:
from plotly.express import scatter
from random import seed
from random import sample

seed(2024)
scatter_df = df[['Item', 'Item Code']].drop_duplicates(ignore_index=True).sort_values(by='Item Code').reset_index(drop=True)
scatter_df['x'] = sample(population=range(1 + len(scatter_df)), k=len(scatter_df))
scatter_df['y'] = sample(population=range(1 + len(scatter_df)), k=len(scatter_df))

scatter(data_frame=scatter_df, x='x', y='y', hover_name='Item', log_y=False, color='Item Code', text='Item', height=900).update_traces(marker={'size': 3})

In [5]:
from plotly.express import histogram
histogram(data_frame=df[['Item', 'Item Code']].drop_duplicates(ignore_index=True).sort_values(by='Item Code').reset_index(drop=True), x='Item Code', y='Item')

In [6]:
df['Item'].value_counts()

Item
Fruit Primary              524
Roots and Tubers, Total    523
Vegetables Primary         522
Vegetables, fresh nes      513
Oilcrops                   510
                          ... 
Quinoa                       8
Cassava leaves               8
Kapok fruit                  6
Gums, natural                3
Jojoba seed                  3
Name: count, Length: 174, dtype: int64

In [7]:
from plotly.express import choropleth
for item in ['Fruit Primary', 'Roots and Tubers, Total', 'Vegetables Primary', 'Vegetables, fresh nes','Oilcrops']:
    choropleth(data_frame=df[(df['Item'] == item) & (df['Element'] == 'Yield')][['Y2019', 'alpha-3',]], locations='alpha-3', color='Y2019',
          title=item).show()